File: test/unit/search.test.ts
import { describe, expect, it } from 'vitest';
import { LogLayout } from '../../src/core/layout/layout.ts';
import { compileSearch, LogSearch } from '../../src/core/layout/search.ts';
import { LogStore } from '../../src/core/store.ts';

const text = (value: string) => ({ parts: [{ type: 'text' as const, text: value }] });

const setup = (lines: string[], maxEntries = 1000) => {
	const store = new LogStore({ maxEntries, mergeRepeats: false });
	const layout = new LogLayout(store);
	const search = new LogSearch(layout);

	store.append(lines.map(text));
	layout.sync();

	return { store, layout, search };
};

const scanAll = (layout: LogLayout, search: LogSearch): void => {
	layout.sync();

	while (search.pending) {
		search.scan(2);
	}
};

describe('compileSearch', () => {
	it('matches the text as typed, ignoring case and special characters', () => {
		const { pattern } = compileSearch('a.b(');

		expect('A.B( axb('.match(pattern!)).toEqual(['A.B(']);
		expect(compileSearch('')).toEqual({ pattern: null, error: null });
	});

	it('matches case and regular expressions when asked, and reports a bad pattern', () => {
		expect('Error error'.match(compileSearch('error', { caseSensitive: true }).pattern!)).toEqual([
			'error'
		]);
		expect('id=12 id=345'.match(compileSearch('id=\\d+', { regex: true }).pattern!)).toEqual([
			'id=12',
			'id=345'
		]);

		const broken = compileSearch('(', { regex: true });

		expect(broken.pattern).toBeNull();
		expect(broken.error).toEqual(expect.any(String));
	});
});

describe('LogSearch', () => {
	it('finds every match as cell ranges, in order, a slice at a time', () => {
		const { layout, search } = setup(['error one', 'nothing', 'Error two, error three']);

		search.setQuery('error');
		expect(search.pending).toBe(true);
		search.scan(1);
		expect(search.count).toBe(1);
		expect(search.pending).toBe(true);
		scanAll(layout, search);

		expect(search.count).toBe(3);
		expect(search.getMatch(1)).toMatchObject({ line: 0, from: 0, to: 5 });
		expect(search.getMatch(2)).toMatchObject({ line: 0, from: 11, to: 16 });
		expect(search.pending).toBe(false);
	});

	it('counts wide characters as two cells and matches decomposed Hangul', () => {
		const { layout, search } = setup([`파일 ${'한글'.normalize('NFD')}.txt`]);

		search.setQuery('한글');
		scanAll(layout, search);

		// `파일 ` takes five cells, and each Hangul syllable two.
		expect(search.getMatch(0)).toMatchObject({ from: 5, to: 9 });
	});

	it('searches again when only the options change, and finds nothing for a bad pattern', () => {
		const { layout, search } = setup(['Error 1', 'error 22']);

		search.setQuery('error');
		scanAll(layout, search);
		expect(search.count).toBe(2);

		expect(search.setQuery('error', { caseSensitive: true })).toBe(true);
		scanAll(layout, search);
		expect(search.count).toBe(1);

		search.setQuery('\\d{2}', { regex: true });
		scanAll(layout, search);
		expect(search.count).toBe(1);
		expect(search.getMatch(0)).toMatchObject({ entryId: 2, from: 6, to: 8 });

		search.setQuery('[', { regex: true });
		expect(search.error).toEqual(expect.any(String));
		expect(search.pending).toBe(false);
		expect(search.count).toBe(0);
	});

	it('moves through the matches and wraps around', () => {
		const { layout, search } = setup(['a', 'a', 'a']);

		search.setQuery('a');
		scanAll(layout, search);

		expect(search.next()?.entryId).toBe(1);
		expect(search.next()?.entryId).toBe(2);
		expect(search.previous()?.entryId).toBe(1);
		expect(search.previous()?.entryId).toBe(3);
		expect(search.next()?.entryId).toBe(1);
	});

	it('searches new entries, forgets dropped ones and keeps the current match', () => {
		const { store, layout, search } = setup(['hit 1', 'hit 2', 'hit 3'], 3);

		search.setQuery('hit');
		scanAll(layout, search);
		search.select(2);

		store.append(text('hit 4'));
		scanAll(layout, search);

		expect(search.count).toBe(3);
		expect(search.getMatch(0)?.entryId).toBe(2);
		expect(search.current).toBe(1);
		expect(search.getMatch(search.current)?.entryId).toBe(3);
	});

	it('starts again when the filter changes and finds the current match again', () => {
		const { layout, search } = setup(['hit one', 'skip', 'hit two']);

		search.setQuery('hit');
		scanAll(layout, search);
		search.select(1);

		layout.setFilter({ text: 'two' });
		scanAll(layout, search);

		expect(search.count).toBe(1);
		expect(search.current).toBe(0);
		expect(search.matchesOf(3)).toHaveLength(1);
		expect(search.matchesOf(1)).toBeUndefined();
	});
});

describe('LogLayout.locatePosition', () => {
	it('returns the wrapped row that shows a cell', () => {
		// Lines never wrap narrower than 16 columns.
		const { layout } = setup(['first', 'aaaa bbbb cccc dddd eeee']);

		layout.setColumns(16);
		layout.sync();

		expect(layout.locatePosition({ entryId: 2, line: 0, cell: 0 })).toEqual({
			entryRow: 0,
			indent: 0
		});
		expect(layout.locatePosition({ entryId: 2, line: 0, cell: 12 })?.entryRow).toBe(0);
		expect(layout.locatePosition({ entryId: 2, line: 0, cell: 20 })?.entryRow).toBe(1);
		expect(layout.locatePosition({ entryId: 9, line: 0, cell: 0 })).toBeNull();
	});
});

File: packages/flutter/test/core/search_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lognal/src/core/filter.dart';
import 'package:lognal/src/core/layout/layout.dart';
import 'package:lognal/src/core/layout/search.dart';
import 'package:lognal/src/core/layout/types.dart';
import 'package:lognal/src/core/store.dart';
import 'package:lognal/src/core/types.dart';

import '../support/text.dart';

LogEntryInit text(String value) => LogEntryInit(parts: <LogPart>[TextPart(value)]);

/// A store, a layout over it and a search over that, as one value.
class SearchContext {
  /// Creates the trio.
  const SearchContext(this.store, this.layout, this.search);

  /// The store the entries went into.
  final LogStore store;

  /// The layout over that store.
  final LogLayout layout;

  /// The search over that layout.
  final LogSearch search;
}

SearchContext setup(List<String> lines, [int maxEntries = 1000]) {
  final LogStore store = LogStore(
    options: LogStoreOptions(maxEntries: maxEntries, mergeRepeats: RepeatMode.keep),
  );
  final LogLayout layout = LogLayout(store);
  final LogSearch search = LogSearch(layout);

  store.append(lines.map(text).toList());
  layout.sync();

  return SearchContext(store, layout, search);
}

void scanAll(LogLayout layout, LogSearch search) {
  layout.sync();

  while (search.pending) {
    search.scan(2);
  }
}

void main() {
  group('compileSearch', () {
    test('matches the text as typed, ignoring case and special characters', () {
      final RegExp pattern = compileSearch('a.b(').pattern!;

      expect(
        pattern.allMatches('A.B( axb(').map((RegExpMatch match) => match.group(0)).toList(),
        <String>['A.B('],
      );
      expect(compileSearch('').pattern, isNull);
      expect(compileSearch('').error, isNull);
    });

    test('matches case and regular expressions when asked, and reports a bad pattern', () {
      expect(
        compileSearch(
          'error',
          const SearchOptions(caseSensitive: true),
        ).pattern!.allMatches('Error error').map((RegExpMatch match) => match.group(0)).toList(),
        <String>['error'],
      );
      expect(
        compileSearch(
          r'id=\d+',
          const SearchOptions(regex: true),
        ).pattern!.allMatches('id=12 id=345').map((RegExpMatch match) => match.group(0)).toList(),
        <String>['id=12', 'id=345'],
      );

      final CompiledSearch broken = compileSearch('(', const SearchOptions(regex: true));

      expect(broken.pattern, isNull);
      expect(broken.error, isA<String>());
    });
  });

  group('LogSearch', () {
    test('finds every match as cell ranges, in order, a slice at a time', () {
      final SearchContext context = setup(<String>[
        'error one',
        'nothing',
        'Error two, error three',
      ]);

      context.search.setQuery('error');
      expect(context.search.pending, isTrue);
      context.search.scan(1);
      expect(context.search.count, 1);
      expect(context.search.pending, isTrue);
      scanAll(context.layout, context.search);

      expect(context.search.count, 3);
      expect(context.search.getMatch(1)?.line, 0);
      expect(context.search.getMatch(1)?.from, 0);
      expect(context.search.getMatch(1)?.to, 5);
      expect(context.search.getMatch(2)?.from, 11);
      expect(context.search.getMatch(2)?.to, 16);
      expect(context.search.pending, isFalse);
    });

    test('counts wide characters as two cells and matches decomposed Hangul', () {
      final SearchContext context = setup(<String>['파일 $decomposedHangul.txt']);

      context.search.setQuery('한글');
      scanAll(context.layout, context.search);

      // `파일 ` takes five cells, and each Hangul syllable two.
      expect(context.search.getMatch(0)?.from, 5);
      expect(context.search.getMatch(0)?.to, 9);
    });

    test('searches again when only the options change, and finds nothing for a bad pattern', () {
      final SearchContext context = setup(<String>['Error 1', 'error 22']);

      context.search.setQuery('error');
      scanAll(context.layout, context.search);
      expect(context.search.count, 2);

      expect(context.search.setQuery('error', const SearchOptions(caseSensitive: true)), isTrue);
      scanAll(context.layout, context.search);
      expect(context.search.count, 1);

      context.search.setQuery(r'\d{2}', const SearchOptions(regex: true));
      scanAll(context.layout, context.search);
      expect(context.search.count, 1);
      expect(context.search.getMatch(0)?.entryId, 2);
      expect(context.search.getMatch(0)?.from, 6);
      expect(context.search.getMatch(0)?.to, 8);

      context.search.setQuery('[', const SearchOptions(regex: true));
      expect(context.search.error, isA<String>());
      expect(context.search.pending, isFalse);
      expect(context.search.count, 0);
    });

    test('moves through the matches and wraps around', () {
      final SearchContext context = setup(<String>['a', 'a', 'a']);

      context.search.setQuery('a');
      scanAll(context.layout, context.search);

      expect(context.search.next()?.entryId, 1);
      expect(context.search.next()?.entryId, 2);
      expect(context.search.previous()?.entryId, 1);
      expect(context.search.previous()?.entryId, 3);
      expect(context.search.next()?.entryId, 1);
    });

    test('searches new entries, forgets dropped ones and keeps the current match', () {
      final SearchContext context = setup(<String>['hit 1', 'hit 2', 'hit 3'], 3);

      context.search.setQuery('hit');
      scanAll(context.layout, context.search);
      context.search.select(2);

      context.store.add(text('hit 4'));
      scanAll(context.layout, context.search);

      expect(context.search.count, 3);
      expect(context.search.getMatch(0)?.entryId, 2);
      expect(context.search.current, 1);
      expect(context.search.getMatch(context.search.current)?.entryId, 3);
    });

    test('starts again when the filter changes and finds the current match again', () {
      final SearchContext context = setup(<String>['hit one', 'skip', 'hit two']);

      context.search.setQuery('hit');
      scanAll(context.layout, context.search);
      context.search.select(1);

      context.layout.setFilter(const LogFilter(text: 'two'));
      scanAll(context.layout, context.search);

      expect(context.search.count, 1);
      expect(context.search.current, 0);
      expect(context.search.matchesOf(3), hasLength(1));
      expect(context.search.matchesOf(1), isNull);
    });
  });

  group('LogLayout.locatePosition', () {
    test('returns the wrapped row that shows a cell', () {
      // Lines never wrap narrower than 16 columns.
      final SearchContext context = setup(<String>['first', 'aaaa bbbb cccc dddd eeee']);

      context.layout.columns = 16;
      context.layout.sync();

      final PositionLocation? start = context.layout.locatePosition(
        const TextPosition(entryId: 2, line: 0, cell: 0),
      );

      expect(start?.entryRow, 0);
      expect(start?.indent, 0);
      expect(
        context.layout.locatePosition(const TextPosition(entryId: 2, line: 0, cell: 12))?.entryRow,
        0,
      );
      expect(
        context.layout.locatePosition(const TextPosition(entryId: 2, line: 0, cell: 20))?.entryRow,
        1,
      );
      expect(
        context.layout.locatePosition(const TextPosition(entryId: 9, line: 0, cell: 0)),
        isNull,
      );
    });
  });
}
